sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -218,7 +247,14 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
       bucketSpec = getBucketSpec,
       options = extraOptions.toMap)
 
-    dataSource.write(mode, df)
+    val destination = source match {
+      case "jdbc" => extraOptions.get(JDBCOptions.JDBC_TABLE_NAME)
+      case _ => extraOptions.get("path")
```

Comment:
But not all of the keys are exposed as public APIs as far as I can see.

 e.g. calling the `save` method adds a "path" key to the option map, but is that key name a public API? I don't consider the key name a public API in this case (you can change it and existing applications will keep working).

Similarly for the JDBC URL and table names. The public method is the API, not necessarily the keys used internally.